pkg/controller/iam/serviceaccountkey.go

```diff
@@ -0,0 +1,336 @@
+		saNamer: RelativeResourceNamer{
+			projectName: projectID,
+		},
+	}
+}
+
+// saKeyrelativeResourceNamer allows the controller to generate the Google Cloud API relative resource name
+// for the service account key based on the external-name annotation.
+// https://cloud.google.com/apis/design/resource_names#relative_resource_name
+// The relative resource names for service account keys have the following format:
+// projects/{project_id}/serviceAccounts/{account}/keys/{key}
+type saKeyrelativeResourceNamer struct {
+	saNamer RelativeResourceNamer
+}
+
+// referencedServiceAccountPath returns the external name of the service account which saKey belongs to
+func (rrn saKeyrelativeResourceNamer) referencedServiceAccountPath(saKey *v1alpha1.ServiceAccountKey) (string, error) {
+	// we assume Google Cloud API relative resource name (aka resource path) of the ServiceAccount is stored in spec.ForProvider.ServiceAccount
+	if saKey.Spec.ForProvider.ServiceAccount == nil || *saKey.Spec.ForProvider.ServiceAccount == "" {
+		return "", fmt.Errorf(fmtErrInvalidServiceAccountRef, saKey.Spec.ForProvider.ServiceAccountReferer)
+	}
+
+	return *saKey.Spec.ForProvider.ServiceAccount, nil
+}
+
+// resourcePath yields the Google Cloud API relative resource name for the ServiceAccountKey resource
+//   returns <the relative resource name>, <whether the external name annotation is non-empty>, <error encountered during resolution>
+func (rrn saKeyrelativeResourceNamer) resourcePath(saKey *v1alpha1.ServiceAccountKey) (string, bool, error) {
+	if saPath, err := rrn.referencedServiceAccountPath(saKey); err != nil {
+		return "", false, err
+	} else if extName := meta.GetExternalName(saKey); extName == "" {
+		return "", false, nil
+	} else {
+		return fmt.Sprintf(fmtKeyRelativeResourceName, saPath, extName), true, nil
+	}
+}
+
+// parseKeyIDFromRrn tries to parse key id from Google Cloud API relative resource name (resource path) of
+//   a service account key
+func parseKeyIDFromRrn(rrn string) (string, error) {
+	resourcePath, err := url.Parse(rrn)
+
+	if err != nil {
+		return "", err
+	}
+
+	return path.Base(resourcePath.Path), nil
+}
+
+func populateSaKey(cr *v1alpha1.ServiceAccountKey, fromProvider *iamv1.ServiceAccountKey) error {
+	keyID, err := parseKeyIDFromRrn(fromProvider.Name)
+
+	if err != nil {
+		return err
+	}
+
+	cr.Status.AtProvider.KeyID = keyID
+	cr.Status.AtProvider.Name = fromProvider.Name
+	cr.Status.AtProvider.PrivateKeyType = fromProvider.PrivateKeyType
+	cr.Status.AtProvider.KeyAlgorithm = fromProvider.KeyAlgorithm
+	cr.Status.AtProvider.ValidAfterTime = fromProvider.ValidAfterTime
+	cr.Status.AtProvider.ValidBeforeTime = fromProvider.ValidBeforeTime
+	cr.Status.AtProvider.KeyOrigin = fromProvider.KeyOrigin
+	cr.Status.AtProvider.KeyType = fromProvider.KeyType
+
+	return nil
+}
+
+func getConnectionDetails(publicKeyType *string, fromProvider *iamv1.ServiceAccountKey) (result managed.ConnectionDetails, err error) {
```

Comment:
It seems this function is used in two places and only one of them executes private key section. What do you think about doing these operations in place? Like `only provided in keys.create responses` in `Create` and `only provided optionally in keys.get responses` in `Observe`.

Reply:
I was saying the following section could go to `Create`:
```
if result[keyPrivateKeyData], err = base64.StdEncoding.DecodeString(fromProvider.PrivateKeyData); err != nil {
  return
}
// only provided in keys.create responses
result[keyPrivateKeyType] = []byte(fromProvider.PrivateKeyType)
```

and the following could go to `Observe`:
```
		if result[keyPublicKeyData], err = base64.StdEncoding.DecodeString(fromProvider.PublicKeyData); err != nil {
			return
		}
		// only provided optionally in keys.get responses
		if publicKeyType != nil {
			result[keyPublicKeyType] = []byte(*publicKeyType)
		}
```

then we'd not need a separate `getConnectionDetails` function.